src/Microsoft.SqlTools.ServiceLayer/QueryStore/QueryStoreService.cs

```diff
@@ -0,0 +1,545 @@
+                    [QueryGeneratorUtils.ParameterPlanId] = requestParams.PlanId,
+                };
+
+                query = PrependSqlParameters(query, sqlParams);
+
+                return new QueryStoreQueryResult()
+                {
+                    Success = true,
+                    ErrorMessage = null,
+                    Query = query
+                };
+            }, requestContext);
+        }
+
+        #endregion
+
+        #endregion
+
+        #region Helpers
+
+        private ColumnInfo GetOrderByColumn(IOrderableQueryParams requestParams, IList<ColumnInfo> columnInfoList)
+        {
+            return requestParams.GetOrderByColumnId() != null ? columnInfoList.First(col => col.GetQueryColumnLabel() == requestParams.GetOrderByColumnId()) : columnInfoList[0];
+        }
+
+        internal virtual IList<Metric> GetAvailableMetrics(QueryStoreReportParams requestParams)
+        {
+            ConnectionService.TryFindConnection(requestParams.ConnectionOwnerUri, out ConnectionInfo connectionInfo);
+
+            if (connectionInfo != null)
+            {
+                using (SqlConnection connection = ConnectionService.OpenSqlConnection(connectionInfo, "QueryStoreService available metrics"))
+                {
+                    return QdsMetadataMapper.GetAvailableMetrics(connection);
+                }
+            }
+            else
+            {
+                throw new InvalidOperationException($"Unable to find connection for '{requestParams.ConnectionOwnerUri}'");
+            }
+        }
+
+        /// <summary>
+        /// Prepends declarations and definitions of <paramref name="sqlParams"/> to <paramref name="query"/>
+        /// </summary>
+        /// <param name="query"></param>
+        /// <param name="sqlParams"></param>
+        /// <returns></returns>
+        private static string PrependSqlParameters(string query, Dictionary<string, object> sqlParams)
+        {
+            StringBuilder sb = new StringBuilder();
+
+            foreach (string key in sqlParams.Keys)
+            {
+                sb.AppendLine($"DECLARE {key} {GetTSqlRepresentation(sqlParams[key])};");
+            }
+
+            sb.AppendLine();
+            sb.AppendLine(query);
+
+            return sb.ToString().Trim();
+        }
+
+        private static HashSet<Type> types = new HashSet<Type>();
+
+        /// <summary>
+        /// Converts an object (that would otherwise be set as a SqlParameter value) to an entirely TSQL representation.
+        /// Only handles the same subset of object types that Query Store query generators use:
+        /// int, long, string, and DateTimeOffset
+        /// </summary>
+        /// <param name="paramValue"></param>
+        /// <returns>data type and value portions of a parameter declaration, in the form "INT = 999"</returns>
+        internal static string GetTSqlRepresentation(object paramValue)
+        {
+            types.Add(paramValue.GetType());
```

Comment:
do the types added to this HashSet get used somewhere later? I couldn't find where when I did a quick search

Reply:
Nope; leftover from making sure I'd collected all the different data types that needed to be handled with this code.